walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,159 @@
+     * the key would be inserted into the list: the index of the first element greater
+     * than the key, or list.size() if all elements in the list are less than the specified key.
+     * This variable signifies where the start time of the meeting fits into the list.
+     */
+    int startIndex = Collections.binarySearch(busyTimes, meetingTime, TimeRange.ORDER_BY_START);
+    // The position where the end time of the meeting fits into the sorted list
+    int endIndex = Collections.binarySearch(busyTimes, meetingTime, TimeRange.ORDER_BY_END);
+    // Finding the actual position from the above value, which may be negative
+    int startPos = (startIndex < 0) ? (-(startIndex + 1)) : startIndex;
+    int prev = startPos - 1;
+    // Finding the nonnegative position of the next element
+    int next = (endIndex < 0) ? (-(endIndex + 1)) : endIndex;
+    ArrayList<TimeRange> newBusyTimes = new ArrayList<>();
+
+    // stores the index in the array from which this meeting begins
+    int startPoint = 0;
+    // stores the time at which this meeting begins
+    int startTime = 0;
+    // stores the index in the array at which this meeting ends
+    int endPoint = 0;
+    // stores the time at which this meeting ends
+    int endTime = 0;
+
+    if (prev < 0) {
+      // If the meeting is to be inserted at the start of the list, startPoint = 0
+      startPoint = startPos;
+      startTime = meetingTime.start();
+    } else {
+      TimeRange prevElem = busyTimes.get(prev);
+      /*
+       * Otherwise find out whether this meeting overlaps with the previous meeting
+       * If it does, coalesce it with the previous meeting otherwise keep them
+       * separate
+       */
+      startPoint = prevElem.overlaps(meetingTime) ? prev : startPos;
+      startTime = prevElem.overlaps(meetingTime) ? prevElem.start() : meetingTime.start();
+    }
+
+    if (next >= busyTimes.size()) {
+      // If this element is to be inserted at the end of the list, endPoint = last element of list
+      endPoint = next - 1;
+      endTime = meetingTime.end();
+    } else {
+      /*
+       * Otherwise find out whether this meeting overlaps with the next meeting
+       * If it does, coalesce it with the next meeting otherwise keep them
+       * separate
+       */
+      TimeRange nextElem = busyTimes.get(next);
+      endPoint = nextElem.overlaps(meetingTime) ? next : next - 1;
+      endTime = nextElem.overlaps(meetingTime) ? nextElem.end() : meetingTime.end();
+    }
+
+    // create a new meeting with the given start and end times
+    TimeRange newBusy = TimeRange.fromStartEnd(startTime, endTime, false);
+
+    /*
+     * This meeting ends before the current first meeting starts, so add it to the
+     * beginning of the list
+     */
+    if (endPoint < 0) {
+      newBusyTimes.add(newBusy);
+    }
+    for (int i = 0; i < busyTimes.size(); i++) {
+      if (i == startPoint && endPoint >= 0 && startPoint < busyTimes.size()) {
```

Comment:
Do you need `startPoint < busyTimes.size()`? If `i == startPoint` then `startPoint < busyTimes.size()` should never be false.

Reply:
Yeah that's true, I was just trying to prevent redundant adds in the special cases addressed before and after the loop but I guess this one never happens :).